patches/0011-implement-ms_tls_config_schannel-experiment.patch

```diff
@@ -0,0 +1,357 @@
+ 9 files changed, 220 insertions(+)
+ create mode 100644 src/crypto/tls/cipher_suites_windows.go
+ create mode 100644 src/crypto/tls/cipher_suites_windows_test.go
+ create mode 100644 src/internal/goexperiment/exp_ms_tls_config_schannel_off.go
+ create mode 100644 src/internal/goexperiment/exp_ms_tls_config_schannel_on.go
+
+diff --git a/src/crypto/tls/cipher_suites_windows.go b/src/crypto/tls/cipher_suites_windows.go
+new file mode 100644
+index 00000000000000..c1dcd97555fe4d
+--- /dev/null
++++ b/src/crypto/tls/cipher_suites_windows.go
+@@ -0,0 +1,74 @@
++// Copyright 2025 The Go Authors. All rights reserved.
++// Use of this source code is governed by a BSD-style
++// license that can be found in the LICENSE file.
++
++//go:build goexperiment.ms_tls_config_schannel
++
++package tls
++
++import (
++	"internal/syscall/windows"
++	"slices"
++	"unsafe"
++)
++
++func init() {
++	cipherSuitesSchannel, err := cipherSuitesSchannel()
++	if err != nil {
++		panic(err) // This should never happen, as Schannel is always available on Windows.
++	}
++	cipherSuitesPreferenceOrder = filterCipherSuites(cipherSuitesSchannel, cipherSuitesPreferenceOrder)
++	cipherSuitesPreferenceOrderNoAES = filterCipherSuites(cipherSuitesSchannel, cipherSuitesPreferenceOrderNoAES)
++	defaultCipherSuitesTLS13 = filterCipherSuites(cipherSuitesSchannel, defaultCipherSuitesTLS13)
++	defaultCipherSuitesTLS13NoAES = filterCipherSuites(cipherSuitesSchannel, defaultCipherSuitesTLS13NoAES)
++	allowedCipherSuitesFIPS = filterCipherSuites(cipherSuitesSchannel, allowedCipherSuitesFIPS)
++	allowedCipherSuitesTLS13FIPS = filterCipherSuites(cipherSuitesSchannel, allowedCipherSuitesTLS13FIPS)
++}
++
++func cipherSuiteID(name string) (uint16, bool) {
++	for _, c := range CipherSuites() {
++		if c.Name == name {
++			return c.ID, true
++		}
++	}
++	for _, c := range InsecureCipherSuites() {
++		if c.Name == name {
++			return c.ID, true
++		}
++	}
++	return 0, false
++}
++
++// cipherSuitesSchannel returns all the cipher suites that Schannel supports in preference order.
++func cipherSuitesSchannel() ([]uint16, error) {
++	// Get all the cipher suites that Schannel supports in preference order.
++	var size uint32
++	var funcs *windows.CRYPT_CONTEXT_FUNCTIONS
++	err := windows.BCryptEnumContextFunctions(windows.CRYPT_LOCAL, unsafe.SliceData(windows.SSL_CONTEXT[:]), windows.NCRYPT_SCHANNEL_INTERFACE, &size, &funcs)
++	if err != nil {
++		return nil, err
++	}
++	defer windows.BCryptFreeBuffer(unsafe.Pointer(funcs))
++
++	suites := make([]uint16, 0, funcs.Count) // order[i] will be the index of suites[i] in the Schannel list
++	for i := range funcs.Count {
++		name := windows.UTF16PtrToString(funcs.At(int(i)))
++		id, ok := cipherSuiteID(name)
++		if !ok {
++			continue // cipher suite not found in the provided list
++		}
++		suites = append(suites, id)
++	}
++	return suites, nil
++}
++
++// filterCipherSuites filters the provided cipher suites, removing those that are not in the allowed list.
```

Comment:
```suggestion
+// filterCipherSuites filters the provided cipher suites, creating a new slice that only includes those that are in the allowed list.
```

With `slices.DeleteFunc` etc., "removing" tends to mean modifying the original.

(I originally added a review comment to potentially use `slices.DeleteFunc`, but now that I've taken a closer look--that would turn out badly.)